### PR TITLE
[FIX] [IMPROVEMENT] keep nft when transaction fail

### PR DESCRIPTION
**Pull request by tommasini**
**Description**
The problem was the collectible was removed when the transaction started.

**Proposed Solution**
The solution proposed was that the collectible was only removed when the transaction was confirmed.
Also, it's proposed here when the transaction it's submitted and we are waiting for feedback from the blockchain, the collectible details will change the _Send_ button for a _View Transactions_ button, which will redirect the user to the Activity View option on the left menu, where he can see the collectible transaction and cancel it or speed it up.

Also sorted the collectible by it's name inside of the collection.

**Code Impact**
Medium, changes the collectible details component and the way that collectibles have in the wallet.

**Screenshots/Recordings**
Small demo of the solution:
UPDATE:  https://recordit.co/6b8oh8xWLG
**Test Cases**
Case1:
 - Import collectible
 - Transact collectible to another account
 - Press the collectible to see the collectible details
 - Click on View Transaction Button, will redirect to Activity View
 - Cancel the collectible transaction 
 - Go to wallet
 - Verify if can send the collectible again
Case2:
 - Import collectible
 - Transact collectible to another account
 - Press the collectible to see the collectible details
 - Wait for the transaction ends and touch the modal to the button update

**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

**https://github.com/MetaMask/metamask-mobile/issues/2784#issuecomment-1038345711**

Solves #2784

### CODE REVIEW

**Comment by chrisleewilcox** [comment]
I don't see the NFT in my receiving account that I had sent to.  Does the NFT need still need to be imported into the receiving account?

**Comment by tommasini** [comment]
@chrisleewilcox , yes you will need to add in the receiving account too, you can also check in the open sea for test networks if you are using one [here](https://testnets.opensea.io/account)

**Review by chrisleewilcox — APPROVED** [review]
LGTM

File: app/components/Views/SendFlow/Confirm/index.js [hunk 1]
```diff
@@ -859,13 +857,20 @@ class Confirm extends PureComponent {
 			if (transactionMeta.error) {
 				throw transactionMeta.error;
 			}
+			if (assetType === 'ERC721') {
```
Comment (gantunesr) [inline app/components/Views/SendFlow/Confirm/index.js:860]: Lets move `'ERC721'` to a constant file and update and the occurrences in the codebase along side `'ERC1155'`

File: app/components/UI/CollectibleOverview/index.js [hunk 2]
```diff
@@ -433,6 +457,14 @@ CollectibleOverview.propTypes = {
 	 * callback to trigger when modal is being animated
 	 */
 	onTranslation: PropTypes.func,
+	/**
+	 * When the collectible is being transacted, this func will navigate to Transactions View
```
Comment (gantunesr) [inline app/components/UI/CollectibleOverview/index.js:461]: Nit

```suggestion
	 * When the collectible is being transacted, this method will navigate to Transactions View
```

File: app/components/UI/CollectibleOverview/index.js [hunk 3]
```diff
@@ -433,6 +457,14 @@ CollectibleOverview.propTypes = {
...
+	 */
+	onViewTransactionDetails: PropTypes.func,
+	/**
+	 * Returns if the collectible is being transacted or not
```
Comment (gantunesr) [inline app/components/UI/CollectibleOverview/index.js:465]: Nit

```suggestion
	 * Indicates if the collectible is being transacted or not
```

**Review by gantunesr — COMMENTED** [review]
Left some minor comments

**Review by gantunesr — CHANGES_REQUESTED** [review]
This PR is blocked by a controllers release

https://github.com/MetaMask/controllers/pull/762

File: app/components/Views/SendFlow/Confirm/index.js (hunk 1 above)
Comment (tommasini) [reply app/components/Views/SendFlow/Confirm/index.js]: Done! 
Created at util/tokens constant file

File: app/components/UI/CollectibleContracts/index.js [hunk 4]
```diff
@@ -104,9 +104,14 @@ const CollectibleContracts = ({
 
   const onItemPress = useCallback(
     (collectible, contractName) => {
-      navigation.navigate('CollectiblesDetails', { collectible, contractName });
+      navigation.navigate('CollectiblesDetails', {
```
Comment (gantunesr) [inline app/components/UI/CollectibleContracts/index.js:107]: Move this constant to the Routes file

File: app/components/UI/CollectibleModal/index.js [hunk 5]
```diff
@@ -35,16 +36,28 @@ const styles = StyleSheet.create({
...
+      chainId,
+    )?.collectible;
 
   const onSend = useCallback(async () => {
     newAssetTransaction({ contractName, ...collectible });
     navigation.replace('SendFlowView');
   }, [contractName, collectible, newAssetTransaction, navigation]);
 
+  const onViewTransactionDetails = useCallback(() => {
+    navigation.navigate('TransactionsHome');
```
Comment (gantunesr) [inline app/components/UI/CollectibleModal/index.js:58]: Move this constant to the Routes file

File: app/components/UI/CollectibleModal/index.js [hunk 6]
```diff
@@ -114,8 +127,10 @@ const CollectibleModal = (props) => {
             collectible={{ ...collectible, contractName }}
             tradable={isTradable()}
             onSend={onSend}
+            onViewTransactionDetails={onViewTransactionDetails}
             openLink={openLink}
             onTranslation={onCollectibleOverviewTranslation}
+            isTransacting={!!collectibleUpdated?.transactionId}
```
Comment (gantunesr) [inline app/components/UI/CollectibleModal/index.js:133]: Could to move this small logic to a method with a name that indicates the purpose?

File: app/components/UI/Transactions/index.js [hunk 7]
```diff
@@ -452,6 +455,24 @@ class Transactions extends PureComponent {
...
+    const resetResult =
+      CollectiblesController.resetCollectibleTransactionStatusByTransactionId(
+        cancelTxId,
+        selectedAddress,
+        chainId,
+      );
+
+    if (!resetResult) {
+      Logger.error({
+        message: `resetCollectibleTransactionStatusByTransactionId failed `,
```
Comment (gantunesr) [inline app/components/UI/Transactions/index.js:471]: Should it be?
```suggestion
        message: 'resetCollectibleTransactionStatusByTransactionId failed',
```

**Review by gantunesr — COMMENTED** [review]
Some small comments

**Review by gantunesr — COMMENTED** [review]
I think there's room for improvement on this PR. I left some small comments, the rest LGTM in the meanwhile

**Review by gantunesr — APPROVED** [review]

File: app/components/Views/SendFlow/Confirm/index.js [hunk 8]
```diff
@@ -747,6 +764,22 @@ class Confirm extends PureComponent {
         navigation && navigation.dangerouslyGetParent()?.pop();
       });
     } catch (error) {
+      Alert.alert(
+        strings('transactions.transaction_error'),
+        error && error.message,
+        [{ text: strings('navigation.ok') }],
+      );
+      Logger.error(error, 'error while trying to send transaction (Confirm)');
```
[the remark this answers is not in the document]
Comment (tommasini) [reply app/components/Views/SendFlow/Confirm/index.js:772]: It was not wise from me to do that, don't knew why I was doing that log!

File: app/util/tokens/index.ts [hunk 9]
```diff
@@ -1,6 +1,8 @@
 /* eslint-disable import/prefer-default-export */
 import { TokenListMap } from '@metamask/controllers';
 
+export const ERC721 = 'ERC721';
```
[the remark this answers is not in the document]
Comment (tommasini) [reply app/util/tokens/index.ts:4]: Moved!

**Comment by tommasini** [comment]
Closing this PR since it's outdated

**Pull request closed without merging**